### PR TITLE
DlgTrackInfo: Use TrackRecord as model

### PR DESCRIPTION
**Pull request by uklotzde**
Save all edits temporarily in a local TrackRecord and replace them in the
Track object all at once on apply. This is required to keep the properties
in sync and valid while editing. Validation should not be done by the
dialog itself but by TrackRecord which serves as the model that contains
the domain logic.

This change becomes crucial when properties dependent on each other,
e.g. when storing genres both as multiple, custom tags and redundantly
formatted in a text property. Those properties need to be synchronized
mutually with each other.

The control and data flow is still far from perfect but it is hopefully a first
step in the right direction.

Should also fix https://bugs.launchpad.net/mixxx/+bug/1929311/comments/18

### CODE REVIEW

**Comment by Holzhaus** [comment]
I suppose this conflicts a bit with #3924, right? Can we return a QFlags when replacing the trackrecord that reflects which values actually changed and then emit changed signals on the track?

**Review by Holzhaus — CHANGES_REQUESTED** [review]
The Track edit dialog acts weird with this PR. When hitting "Apply", values disappear: 
Peek 2021-05-29 18-52

**Comment by uklotzde** [comment]
@Holzhaus Thanks for testing! Editing the genre field works differently with custom tags. I forgot to add the signal handler that is needed temporarily until custom tags are ready.

I also found another minor initialization bug in the opposite direction when populating fields. Rust would have saved me from making such an obvious mistake ;)

**Comment by uklotzde** [comment]
> I suppose this conflicts a bit with #3924, right? Can we return a QFlags when replacing the trackrecord that reflects which values actually changed and then emit changed signals on the track?

Unrelated. The flags would need to be passed internally to Track::emitMetadataChanged() within `Track`.

**Comment by Holzhaus** [comment]
If I change the Genre, the key disappears:

Peek 2021-06-09 12-41

File: src/library/dlgtrackinfo.cpp [hunk 1]
```diff
@@ -384,54 +479,29 @@ void DlgTrackInfo::saveTrack() {
...
-
-    // Reconnect changed signals now.
-    connect(m_pLoadedTrack.get(),
-            &Track::changed,
-            this,
-            &DlgTrackInfo::slotTrackChanged);
-}
+    // Special case handling for the comment field that is not
+    // updated by the editingFinished signal.
+    m_trackRecord.refMetadata().refTrackInfo().setComment(txtComment->toPlainText());
```
Comment (Holzhaus) [inline src/library/dlgtrackinfo.cpp:486]: Why is it not updated by the editingFinished signal?

**Review by Holzhaus — APPROVED** [review]
LGTM. Code changes seem sensible.

**Comment by uklotzde** [comment]
@Holzhaus I force pushed the last commit, which was not finished yet. 

The changes not only fix the key text synchronization bug but also remove redundant code and improve the data binding with `TrackRecord`. Only done for the key field, yet. With the introduction of QML we need to rethink all data bindings anyway and come up with consistent patterns and naming.

**Comment by Holzhaus** [comment]
> With the introduction of QML we need to rethink all data bindings anyway and come up with consistent patterns and naming.

With our current QML implementation we can keep using legacy widgets such as the DlgTrackInfo side by side. In the long-term: Yes.

File: src/library/dlgtrackinfo.cpp (hunk 1 above)
Comment (uklotzde) [reply src/library/dlgtrackinfo.cpp:486]: This has not changed. Because QPlainTextEdit only has a textChanged() but no editingFinished() signal?

**Comment by uklotzde** [comment]
This was also a perfect example for marking a data type as `[[nodiscard]]`.

File: src/library/dlgtrackinfo.cpp (hunk 1 above)
Comment (Holzhaus) [reply src/library/dlgtrackinfo.cpp:486]: OK, I was just wondering.

**Pull request merged**